### DISCUSSION TITLE
Waldorf Pulse 2 Arpeggiator Pattern

### DISCUSSION THREAD

**Discussion by synthmania67** [RomanKubiak/ctrlr#554] [General]
Hi,

I´m trying to build a panel only for the Pulse 2 Arpeggiator Pattern, because I want to control all 16 Steps via Cubase.

The manual of the sysex implementation shows me values which I do not understand. For the 16 Events i use a combo with all the necessary names. The sysex code I put in is "F0 3E 16 00 20 5a ls F7" for the first step and it works, but also the time event is set always to "25%". This should not happen. Therefor I want to use also a combo with the 4 possible values.
And also a button for switching on/off the glide per event.

I´ve marked the description at the end of this document. 

**Maybe someone can help me building the right SYSEX message for the combo I want to use.**

[Pulse_2_Sysex_&_CC_Parameter_List_V1.14.pdf](https://github.com/RomanKubiak/ctrlr/files/11520767/Pulse_2_Sysex_._CC_Parameter_List_V1.14.pdf)

![image](https://github.com/RomanKubiak/ctrlr/assets/110454569/490927e2-6dc6-4071-9fe7-d8588e2a220f)
![image](https://github.com/RomanKubiak/ctrlr/assets/110454569/b7835bfe-3b30-45de-a38d-568c9e072416)
![pulse2-display](https://github.com/RomanKubiak/ctrlr/assets/110454569/37516613-7594-4cd1-8f3c-adc24fa9a068)

**Comment by dnaldoog** [comment]
Hi @synthmania67 

You need to populate the top three bits with data, otherwise you are always sending 0 (25%) for the step duration and 0 for (Slide). 

The coding is best done in lua.

With everything set to max, the value sent should be `0x78` or `120`. 🤔

The solution is to combine the three bit packed values and bit shift them up to the appropriate bit position: 
You can optionally use the bit library to do this e.g: ` ll = bit.lshift(jmax(0, ll),4)`,`g0= bit.lshift(g0,6)`

Of course I can't test this but it should be about right. See panel and code: 

[Waldorf Pulse 2 Arpeggiator Pattern_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/11521912/Waldorf.Pulse.2.Arpeggiator.Pattern_1_0.zip)

![waldorf arpeggiator](https://github.com/RomanKubiak/ctrlr/assets/76831683/1784a247-99b2-4b06-b624-07bfd9110d28)

<pre>
send_arp_pattern_format = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    --0glltttt (7 bits of data) 8th bit never used for MIDI data

    local tttt = panel:getModulatorByName("Type of step"):getModulatorValue()
    local ll = panel:getModulatorByName("Length of step"):getModulatorValue()
    local g0 = panel:getModulatorByName("Step glide"):getModulatorValue()

    tttt = jmax(0, tttt) -- an unset combo = -1 so ensure lowest value = 0
    ll = jmax(0, ll) * 16 -- an unset combo = -1 // bit shift left by 4
    g0 = g0 * 64 -- bit shift left by 6
    local str = string.format("F0 3E 16 00 20 5A %.2X F7", tttt + ll + g0)
    panel:sendMidiMessageNow(CtrlrMidiMessage(str))
end
</pre>

**Reply by synthmania67** [reply]
Hi John, I don´t know what to say now. You´re totaly genius, because your code does the trick.  I´m looking for a solution since one week or maybe more.
Thank you very much indeed for your help and kind regards,
synthmania67

**Reply by dnaldoog** [reply]
Excellent to hear!
